File: leetcode/0002-add-two-numbers/solution.go
/**
 * Author: 1chooo<hugo970217@gmail.com>
 * Problem: https://leetcode.com/problems/add-two-numbers/
 * Runtime: 0ms (100.00%)
 */

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := &ListNode{}
	dummyHead := dummy
	carry := 0

	for l1 != nil && l2 != nil {
		sum := l1.Val + l2.Val + carry
		dummy.Next = &ListNode{Val: sum % 10}
		carry = sum / 10

		l1 = l1.Next
		l2 = l2.Next
		dummy = dummy.Next
	}

	if l1 == nil {
		l1, l2 = l2, l1
	}

	for l1 != nil {
		sum := l1.Val + carry
		dummy.Next = &ListNode{Val: sum % 10}
		carry = sum / 10

		l1 = l1.Next
		dummy = dummy.Next
	}

	if carry != 0 {
		dummy.Next = &ListNode{Val: carry}
	}

	return dummyHead.Next
}

File: leetcode/0002-add-two-numbers/solution.py
"""
* Author: 1chooo<hugo970217@gmail.com>
* Problem: https://leetcode.com/problems/add-two-numbers/
* Runtime: 3ms (70.91%)
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = ListNode()
        current = dummy 
        carry = 0
        
        while l1 or l2 or carry:
            sum_val = carry
            
            if l1:
                sum_val += l1.val
                l1 = l1.next
            if l2:
                sum_val += l2.val
                l2 = l2.next
            
            carry, val = divmod(sum_val, 10)
            current.next = ListNode(val)
            current = current.next
        
        return dummy.next
